nesta/core/routines/arxiv/magrxiv_collect_iterative_task.py

```diff
@@ -15,6 +16,74 @@
 from nesta.core.luigihacks import misctools
 from nesta.core.luigihacks.mysqldb import MySqlTarget
 
+def get_article_ids(session, xivs):
+    """Get the set of all article IDs in the current database.
+
+    Args:
+        session (SqlAlchemy connectable): SqlAlchemy db session for querying
+        xivs (list): List of valid 'article_source' values to include in the filtering.
+    Returns:
+        ids (set): set of all article IDs in the current database for these article_sources.
+    """
+    xiv_filter = Article.article_source.in_(xivs)
+    return {article.id for article in
+            session.query(Article.id).filter(xiv_filter).all()}
+
+
+def get_articles(xivs, api_key, start_date, article_ids, test, flush_every=1000):
+    """Get all articles from MAG from specified 'Journals' (xivs).
+
+    Args:
+        xivs (list): List of valid MAG 'Journals' to query.
+        api_key (str): MAG API key
+        start_date (str): Sensibly formatted date string (interpretted by pd)
+        article_ids (set): Set of article IDs to skip. NB: the set will be updated.
+        test (bool): Running in test mode?
+        flush_every (int): Sets the number of articles retrieved between log messages,
+                           and also related to the maximum number of articles acquired
+                           in test mode (NB: in test mode at most 2 x flush_every will be collected).
+    Returns:
+        articles (:obj:`list` of :obj:`dict`): List of articles from MAG, formatted a-la-arxiv.
+    """
+    logging.info(f"{len(article_ids)} existing articles will not be recollected")
+    articles = []
+    for xiv in xivs:
+        articles += _get_articles(xiv, api_key, start_date, article_ids, test, flush_every)
```

Comment:
Rather than hidden mutation of `article_ids` could do something along the lines of the following:

```python
# 1
import operator as op
article_ids.update([op.getitem(row, 'id') for row in articles])
# 2
from toolz.curried import get_in
article_ids.update(map(get_in('id'), articles)
```

`_get_articles` could then be a comprehension (if it wasn't for the logging)

The logging could actually just be a re-usable generator function that wraps an iterable, something like:
 ```python

def foo(x, flush_every, test=True): 
     counter = 0 
     x = iter(x) 
     for xi in x: 
         counter += 1 
         if counter % flush_every == 0: 
             logging.info(f'Processed {counter} items') 
             if test:
                 return xi 
         yield xi 

# in body of `get_articles`:
if test:
    _get_articles = foo(_get_articles)
for xiv in xivs:
    articles = [row for row in get_magrxiv_articles(xiv, api_key, start_date=date) if row['id'] not in article_ids]
    article_ids.update(map(get_in('id'), articles)
```

Reply:
Actually, this solution doesn't work :facepalm: the mutation is one level above still

Reply:
Yep, good shout. Creating a copy of `article_ids` isn't a big issue, and then the ids can just be updated as you suggest.